📝 Update writing tests page (#4116)
reduxjs/redux: docs/recipes/WritingTests.md

# Writing Tests

> 

Reviewer: :hand: If we're updating this page, let's add a "What You'll Learn" `tip` block here similar to what's in the tutorial pages

Author: Yup, added.

I couldn't quite think of anything to fill it with besides "Recommended practices for testing apps using Redux"

Because most of the Redux code you write are functions, and many of them are pure, they are easy to test without mocking. However, you should consider whether each piece of your Redux code needs it's own dedicated tests.

> :::tip

Reviewer: :hand: Let's make boxes that contain links into `info` blocks instead of `tip` blocks

Author: Done

:::

### Guiding Principles

The guiding principles for testing Redux logic closely follow that of React Testing Library:

> [The more your tests resemble the way your software is used, the more confidence they can give you.](https://twitter.com/kentcdodds/status/977018512689455106) - Kent C. Dodds

In the majority of scenarios, the end-user does not know, and does not care whether Redux is used within the application at all.
As such, the Redux code can be treated as an implementation detail of the app, without requiring explicit tests for the Redux code in many circumstances.

The general advice for testing an app using Redux is as follows:

- Use integration tests for everything working together. I.e. for a React app using Redux, render a `<Provider>` with a real store instance wrapping the component/s being tested. Interactions with the page being tested should use real Redux logic, with API calls mocked out so app code doesn't have to change, and assert that the UI is updated appropriately.

> - Use basic unit tests _where fitting_ for pure functions that deserve it. I.e. particularly complex reducers or selectors. However, in many cases, these are just implementation details that are covered by integration tests instead.

Reviewer: :hand: Let's rephrase this:

```suggestion
- If needed, use basic unit tests for pure functions such as particularly complex reducers or selectors. However, in many cases, these are just implementation details that are covered by integration tests instead.
```

:::tip

See [Testing Implementation Details](https://kentcdodds.com/blog/testing-implementation-details) for Kent C. Dodds' thoughts on why he recommends avoiding testing implementation details.

:::

### Setting Up

> We recommend [Jest](https://facebook.github.io/jest/) as the testing engine.

Reviewer: :hand: Did we already have the Jest recommendation here?

I'd like to reword this section to emphasize that Redux can be tested with any test runner, but the examples will use Jest as that's common.

Author: Yeah, that bit was existing (I just added a line about `jsdom`). I re-worded it more in line with your suggestion

and run `npm test` to run it once, or `npm run test:watch` to test on every file change.

> ### Action Creators & Async Action Creators

Reviewer: :hand: I'd prefer _not_ to use the phrase "async action creators" - I've been trying to drop that terminology from the docs.  Let's call them "Thunks" instead

Author: Done, including an instance of "created async thunk action creator" I had added 😅

> In Redux, action creators are functions which return plain objects. Our recommendation is not to write action creators manually, but instead have them generated automatically by [`createSlice`](https://redux-toolkit.js.org/api/createSlice#return-value), or created via [`createAction`](https://redux-toolkit.js.org/api/createAction) from [`@reduxjs/toolkit`](https://redux-toolkit.js.org/introduction/getting-started). As such, you should not feel the need to test action creators by themselves (the Redux Toolkit maintainers have already done that for you!).

Reviewer: 👍

We consider thunk behavior to be an implementation detail of the application, and recommend that it be covered by testing the group of components (or whole app) using it, rather than testing the thunk in isolation.

Our recommendation is to mock async requests at the `fetch/xhr` level using tools like [`msw`](https://mswjs.io/), [`miragejs`](https://miragejs.com/), [`jest-fetch-mock`](https://github.com/jefflau/jest-fetch-mock#readme), [`fetch-mock`](https://www.wheresrhys.co.uk/fetch-mock/), or similar. By mocking requests at this level, none of the thunk logic has to change in a test - the thunk still tries to make a "real" async request, it just gets intercepted. See the [components example](#example-1) for an example of testing a component which internally includes the behavior of async action creators.

### Reducers

> A reducer should be a pure function that returns the new state after applying the action to the previous state. In the majority of cases, the reducer is an implementation detail that does not need explicit tests. However, if your reducer contains particularly complex logic that you would like the confidence of having unit tests for, reducers can be easily tested. As a reducer should always be a pure function, it is easy to write tests for - a given input should always provide a particular output.

Reviewer: ```suggestion
Reducers pure functions that return the new state after applying the action to the previous state. In the majority of cases, the reducer is an implementation detail that does not need explicit tests. However, if your reducer contains particularly complex logic that you would like the confidence of having unit tests for, reducers can be easily tested. 

Because reducers are pure functions, testing them should be straightforward.  Call the reducer with a specific input `state` and `action`, and assert that the result state matches expectations.
```

    status: 'idle'
  },
  reducers: {},
  extraReducers: builder => {
    builder.addCase(fetchUser.pending, (state, action) => {
      state.status = 'loading'
    })
    builder.addCase(fetchUser.fulfilled, (state, action) => {
      state.status = 'complete'
      state.name = action.payload
    })
  }
})

export const selectUser = state => state.user.name
export const selectUserFetchStatus = state => state.user.status

export default userSlice.reducer
```

```jsx
import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { fetchUser, selectUser, selectUserFetchStatus } from './userSlice'

export default function App() {
  const dispatch = useDispatch()
  const user = useSelector(selectUser)
  const userFetchStatus = useSelector(selectUserFetchStatus)

  return (
    <div>
      {/* Display the current user name */}
      <div>{user}</div>
      {/* On button click, dispatch a thunk action to fetch a user */}
      <button onClick={() => dispatch(fetchUser())}>Fetch user</button>
      {/* At any point if we're fetching a user, display that on the UI */}
      {userFetchStatus === 'loading' && <div>Fetching user...</div>}
    </div>
  )
}
```

> This app involves async action creators, reducers and selectors. All of these can be tested by writing an integration test with the following in mind:

Reviewer: :hand 

```suggestion
This app involves thunks, reducers and selectors. All of these can be tested by writing an integration test with the following in mind:
```
